### PR TITLE
Add rewind depth support for execLocal

### PR DESCRIPTION
**Pull request by ak3n**
This PR fixes the behaviour of `execLocal` functionality with a provided rewind depth parameter — it executes the transaction in the context of the block with the provided depth.

### CODE REVIEW

File: src/Chainweb/Pact/PactService.hs
Comment (ak3n) [inline src/Chainweb/Pact/PactService.hs:661]: I'm not sure about renaming the "rewind depth" to the "confirmation depth". Because the semantics is that we want to rewind the block, not to confirm something. At least that's what I have from user's POV.

File: test/Chainweb/Test/Pact/PactMultiChainTest.hs
Comment (ak3n) [inline test/Chainweb/Test/Pact/PactMultiChainTest.hs:237]: Maybe there should be more testing scenarios? When the depth is too big or negative for example? What do you think?

Comment (jwiegley) [reply test/Chainweb/Test/Pact/PactMultiChainTest.hs:237]: I think that's a great idea.

File: src/Chainweb/Pact/PactService.hs
Comment (jwiegley) [reply src/Chainweb/Pact/PactService.hs:661]: This makes sense to me. Rewind is the action, confirmation is an interpretation.

Comment (emilypi) [reply src/Chainweb/Pact/PactService.hs:661]: +1 to both of u

File: test/Chainweb/Test/Pact/PactMultiChainTest.hs
Comment (emilypi) [reply test/Chainweb/Test/Pact/PactMultiChainTest.hs:237]: @ak3n Let's make sure we have coverage for negative rewind depths, and consult @larskuhtz on what a reasonable maximum rewind depth would be good for the nodes. I would assume we don't want people walking all the way back to genesis.

**Review by edmundnoble — APPROVED** [review]
Approved pending selection of a maximum rewind limit. My two cents: 100 is a fine number. Exchanges may have an opinion here.

File: src/Chainweb/Pact/PactService.hs
Comment (larskuhtz) [reply src/Chainweb/Pact/PactService.hs:661]: I think, on the external API "confirmation" is what the user wants. Internally, it's rewind.

So, user facing should be "confirmation depth" (it's a common term in Bitcoin). Rewind is an implementation details of the our checkpointer which has no meaning to external users.

File: test/Chainweb/Test/Pact/PactMultiChainTest.hs
Comment (larskuhtz) [reply test/Chainweb/Test/Pact/PactMultiChainTest.hs:237]: In particular we have to test interactions with the different rewind limits that we have in place.

Generally, `local` is used via the service API, so we can be generous. We should give the user the ability to configure that, because it's their call how much load they expect and how much they are willing to compromise and latencies for other pact requires.

In practice, I expect that more than a few hundred blocks (also depending on the chain) would probably cause significant slow down. So, we may either put a limit or warn the user if they configure a larger number.

Comment (larskuhtz) [reply test/Chainweb/Test/Pact/PactMultiChainTest.hs:237]: This is easy to test, because we log execution time traces for all pact calls. So we can run a node with a large limit and just plot latencies pact request latencies (also of other pact requests) over the requested confirmation depth.

**Review by larskuhtz — COMMENTED** [review]
Looks good to me. I think, what is missing is 

*  some more tests for larger rewind depth in order to test possible interactions with existing rewind limits,

*  addition of a configuration option for setting the maximum allowed local rewind/confirmation depth ("confirmation depth" might be the better  term here since it is user facing), and

* benchmarking (possibly manual, e.g. by deploying to testnet or mainnet node) to determine a reasonable default limit.

If testing large depth is not feasible in unit tests, it could possibly be done on devnet.

File: src/Chainweb/Pact/PactService.hs
Comment (ak3n) [reply src/Chainweb/Pact/PactService.hs:661]: I've got some doubts about the "rewind depth" because it clashes with the checkpointer's rewind mechanism. We will get two different rewind semantics — the local command's rewind of blocks and checkpointer's rewind (the first uses the latter's).

We could say that the checkpointer stuff is an internal thing and users shouldn't know about it, but it's already leaked as an error message, see `PactException(RewindLimitExceeded)`. Also there is `reorgLimit` in `execValidateBlock`, which is a limit for checkpointer's rewind but with additional semantic.

Do we want our users to know about the checkpointer and how much? If not, then there will be two rewinds — 
"rewind depth" as an external API thing and "checkpointer rewinds" as an internal meaning, which we can try to hide. If yes, then we should be more clear and accurate with the terms to avoid ambiguity (still not a fan of "confirmation depth" in this context).

File: test/Chainweb/Test/Pact/PactMultiChainTest.hs
Comment (ak3n) [inline test/Chainweb/Test/Pact/PactMultiChainTest.hs:270]: Two tests — with a negative depth, which tests the limit, and a depth reaching the genesis block.

**Comment by ak3n** [comment]
Some notes:

1. For some reason macos build fails with a dns error

```
dieVerbatim: user error (Error: cabal: '/usr/local/opt/curl/bin/curl' exited with an error:
curl: (6) Could not resolve host: objects-us-east-1.dream.io
```

2. I've discovered a bug with the checkpointer, right now `local` command changes the checkpointer's state and we should bracket it, rewind it back after running the transaction.

File: src/Chainweb/Pact/PactService.hs
Comment (larskuhtz) [reply src/Chainweb/Pact/PactService.hs:665]: I think, `confirmation depth` is a notion only at service REST API level.  "rewind depth" is a concept at checkpoint API level.

The idea is that in the future all service API endpoints that retrieve information from the blockchain would have a common "confirmation depth" parameter. This would include latest cut, querying headers, querying payloads, local calls, and possibly more.

"Rewind" wouldn't be a meaningful notion at the service API level, because it is relative to the checkpoint state which is hidden from the service API user. It also would not be meaningful with respect to querying cuts, headers, or payloads. Instead what service API users care about is the "level of confirmation" of the retrieved data, i.e. the depth in the chain with respect to the current (local) state of consensus.

Comment (larskuhtz) [reply src/Chainweb/Pact/PactService.hs:665]: I wonder if we need to define more precisely what the notion of "current block" means within the context of the checkpointer. Is it just the current state of the checkpoint, or is the the latest state of consensus (which is external to the checkpointer). Clearly both are related, but the respective invariants are currently not very well documented or reflected in the code.

File: src/Chainweb/Pact/Types.hs
Comment (ak3n) [reply src/Chainweb/Pact/Types.hs]: I've raised the `defaultLocalRewindDepthLimit` to `1000`, the usual response time with `1000` is about 5 seconds.

File: test/Chainweb/Test/Pact/PactMultiChainTest.hs
Comment (emilypi) [inline test/Chainweb/Test/Pact/PactMultiChainTest.hs:283]: In terms of style points here, I think we can agree to put a nix on explicit braces and choose good formatting instead. Suggesting: 

```haskell
  handle 
    (\case
        LocalRewindGenesisExceeded -> return ()
        err -> liftIO $ assertFailure $ "Expected LocalRewindGenesisExceeded, but got " ++ show err)
    (liftIO $ do
      r <- runLocalWithDepth (Just $ BlockHeight 56) cid getSender00Balance
      print r
      assertFailure "Expected LocalRewindGenesisExceeded, but block succeeded")

Comment (jwiegley) [reply test/Chainweb/Test/Pact/PactMultiChainTest.hs:283]: Yes please to no explicit braces.

Comment (ak3n) [reply test/Chainweb/Test/Pact/PactMultiChainTest.hs]: Fixed, thanks!

**Comment by ak3n** [comment]
Ok, looks like I have fixed everything. The last and the only thing is how to properly name it. And if it's even possible at this moment? If it's already a part of API as `rewindDepth`?

**Comment by larskuhtz** [comment]
> Ok, looks like I have fixed everything. The last and the only thing is how to properly name it. And if it's even possible at this moment? If it's already a part of API as `rewindDepth`?

Is the depth parameter currently exposed in the API? If not we may want to add it (otherwise the feature wouldn't be available to users). Otherwise, would we have to provide backward compatibility? I think, we never announced it, so we should be fine to break it if we want.

File: src/Chainweb/Pact/Service/Types.hs
Comment (edmundnoble) [inline src/Chainweb/Pact/Service/Types.hs:175]: Nit, I think either both of these should be `BlockHeight` or both of these should be `Natural`, because they're both the same "type of thing" (block height delta)

File: src/Chainweb/Pact/Types.hs
Comment (edmundnoble) [inline src/Chainweb/Pact/Types.hs:360]: There's no need to `UNPACK` this or any other word-sized field anymore, `-funbox-small-strict-fields` is the default.

File: src/Chainweb/Pact/PactService.hs
Comment (edmundnoble) [inline src/Chainweb/Pact/PactService.hs:673]: Maybe we want to throw or log an error here if `seekAncestor` returns `Nothing`, I don't see any legal situation where that can happen.

File: src/Chainweb/Pact/Service/Types.hs
Comment (ak3n) [reply src/Chainweb/Pact/Service/Types.hs:175]: I'm being consistent here with `RewindLimitExceeded`, where the limit is `Natural` and heights are `BlockHeight`. 

And semantics is a bit different — limit says how far in the past we can go, while the height starts from the bottom. Different directions.

File: src/Chainweb/Pact/Types.hs
Comment (ak3n) [reply src/Chainweb/Pact/Types.hs:360]: Good point 👍 , added because of `_psReorgLimit`, will remove for both of them.

File: src/Chainweb/Pact/PactService.hs
Comment (ak3n) [reply src/Chainweb/Pact/PactService.hs:673]: Yeah, that behavior is caught by throwing `LocalRewindGenesisExceeded` above, maybe should replace the check with just throwing an exception here.

Comment (ak3n) [reply src/Chainweb/Pact/PactService.hs]: Fixed by throwing the `BlockHeaderLookupFailure` exception because it's indeed an exceptional situation and shouldn't happen.

File: src/Chainweb/Pact/Service/Types.hs
Comment (edmundnoble) [reply src/Chainweb/Pact/Service/Types.hs:175]: Aren't they both measured from the tip of the chain?

Comment (ak3n) [reply src/Chainweb/Pact/Service/Types.hs:175]: The rewind's limit starts from the current moment, how far we can go backwards.

**Review by edmundnoble — APPROVED** [review]

**Review by chessai — APPROVED** [review]

**Pull request merged**
